plugins/kotlin/idea/src/org/jetbrains/kotlin/idea/inspections/ReplaceManualRangeWithIndicesCallsInspection.kt

```diff
@@ -42,29 +39,20 @@ class ReplaceManualRangeWithIndicesCallsInspection : AbstractKotlinInspection()
         }
     }
 
-    private fun visitRange(holder: ProblemsHolder, expression: KtExpression, left: KtExpression, right: KtExpression, method: String) {
+    private fun visitRange(holder: ProblemsHolder, range: KtExpression, left: KtExpression, right: KtExpression, rangeFunction: String) {
         if (left.toIntConstant() != 0) return
-        val collection = when(method) {
-            "until" -> right.receiverIfIsSizeOrLengthCall()
-            "rangeTo" -> right.receiverIfIsSizeOrLengthMinusOneCall()
-            else -> null
-        } as? KtSimpleNameExpression ?: return
-        if (!collection.doesHaveIndicesExtensionFunction()) return
-        visitIndicesRange(holder, expression, collection)
-    }
-
-    private fun KtExpression.doesHaveIndicesExtensionFunction() =
-        getType(analyze(BodyResolveMode.PARTIAL))?.isMap(DefaultBuiltIns.Instance) == false
+        val sizeOrLengthCall = right.sizeOrLengthCall(rangeFunction) ?: return
+        val collection = (sizeOrLengthCall as? KtQualifiedExpression)?.receiverExpression
```

Comment:
minor: there is quite useful shortcut `safeAs< KtQualifiedExpression>()` that works as `as ?`. In this  case brackets `( ... as? ... )` becomes redundant.